// @sudoku/stores/domainGame.js
import { derived, writable } from 'svelte/store'
import { decodeSencode, encodeSudoku } from '@sudoku/sencode'
import { generateSudoku, solveSudoku } from '@sudoku/sudoku'
import { createGame, createSudoku } from '../../../domain/index.js'
import { hints } from './hints'

function cloneGrid(grid) {
	return grid.map((row) => row.slice())
}

function emptyGrid() {
	return Array.from({ length: 9 }, () => Array(9).fill(0))
}

function safeSolve(grid) {
	try {
		return solveSudoku(grid)
	} catch (error) {
		return null
	}
}

function createGameStore() {
	const view = writable(null)

	let currentGame = null
	let lastHint = null

	function snapshot() {
		const initialGrid = currentGame.getInitialGrid()
		const currentGrid = currentGame.getSudoku().getGrid()
		const exploreStatus = currentGame.getExploreStatus()

		return {
			initialGrid,
			currentGrid,
			invalidCellKeys: currentGame
				.validate()
				.map(({ row, col }) => `${col},${row}`),
			canUndo: currentGame.canUndo(),
			canRedo: currentGame.canRedo(),
			won: currentGame.isWon(),
			hint: lastHint,
			explore: exploreStatus,
		}
	}

	function sync() {
		view.set(snapshot())
	}

	function replaceGame(nextGame) {
		currentGame = nextGame
		lastHint = null
		sync()
	}

	function bootstrap(grid) {
		replaceGame(createGame({
			sudoku: createSudoku(grid),
			solutionGrid: safeSolve(grid),
		}))
	}

	bootstrap(emptyGrid())

	return {
		subscribe: view.subscribe,

		startGenerated(difficulty) {
			bootstrap(generateSudoku(difficulty))
		},

		startFromSencode(sencode) {
			bootstrap(decodeSencode(sencode))
		},

		guess(position, value) {
			if (!position || position.x === null || position.y === null) return false
			const changed = currentGame.guess({ row: position.y, col: position.x, value })
			if (changed) {
				lastHint = null
				sync()
			}
			return changed
		},

		getCandidateHint(position) {
			if (!position || position.x === null || position.y === null) return null
			lastHint = currentGame.getCandidateHint({ row: position.y, col: position.x })
			sync()
			return lastHint
		},

		getNextHint() {
			lastHint = currentGame.getNextHint()
			sync()
			return lastHint
		},

		applyHint(position = null) {
			const domainPosition = position && position.x !== null && position.y !== null
				? { row: position.y, col: position.x }
				: null

			const changed = currentGame.applyHint(domainPosition)
			lastHint = currentGame.getNextHint()

			if (changed) {
				hints.useHint()
			}

			sync()
			return changed
		},

		undo() {
			if (currentGame.undo()) {
				lastHint = null
				sync()
				return true
			}
			return false
		},

		redo() {
			if (currentGame.redo()) {
				lastHint = null
				sync()
				return true
			}
			return false
		},

		startExplore() {
			const changed = currentGame.startExplore()
			if (changed) sync()
			return changed
		},

		commitExplore() {
			const changed = currentGame.commitExplore()
			if (changed) {
				lastHint = null
				sync()
			}
			return changed
		},

		abandonExplore() {
			const changed = currentGame.abandonExplore()
			if (changed) {
				lastHint = null
				sync()
			}
			return changed
		},

		markExploreFailed() {
			const changed = currentGame.markExploreFailed()
			if (changed) sync()
			return changed
		},

		getGame() {
			return currentGame
		},

		getPuzzleGrid() {
			return cloneGrid(currentGame.getInitialGrid())
		},

		encodeCurrentPuzzle() {
			return encodeSudoku(currentGame.getInitialGrid())
		},
	}
}

export const gameStore = createGameStore()

export const puzzleGrid = derived(gameStore, ($gameStore) => $gameStore.initialGrid)
export const currentGrid = derived(gameStore, ($gameStore) => $gameStore.currentGrid)
export const invalidCells = derived(gameStore, ($gameStore) => $gameStore.invalidCellKeys, [])
export const canUndo = derived(gameStore, ($gameStore) => $gameStore.canUndo, false)
export const canRedo = derived(gameStore, ($gameStore) => $gameStore.canRedo, false)
export const gameWon = derived(gameStore, ($gameStore) => $gameStore.won, false)
export const currentHint = derived(gameStore, ($gameStore) => $gameStore.hint, null)
export const exploreStatus = derived(gameStore, ($gameStore) => $gameStore.explore, {
	active: false,
	conflict: false,
	repeatedFailure: false,
	failedPathCount: 0,
})
export const exploring = derived(exploreStatus, ($exploreStatus) => $exploreStatus.active, false)
